### PR TITLE
Add solutions for advent of code 2024 day 1 to `examples/`.

### PR DESCRIPTION
**Pull request by zygoloid**
In order to support these examples, this adds two new builtins to the toolchain: `print.char` and `read.char`, which map to the libc functions `putchar` and `getchar`.

### CODE REVIEW

**Comment by zygoloid** [comment]
Depends on #4644.

File: examples/advent2024/day1_part2.carbon [hunk 1]
```diff
@@ -0,0 +1,149 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+library "day1_common";
+
+var EOF: i32 = -(1 as i32);
+var push_back: i32 = EOF;
+
+fn ReadCharRaw() -> i32 = "read.char";
```
Comment (jonmeow) [inline examples/advent2024/day1_part2.carbon]: Can these be moved out of `examples/` and into a library?

File: toolchain/check/eval.cpp [hunk 2]
```diff
@@ -1107,8 +1107,11 @@ static auto MakeConstantForBuiltinCall(Context& context, SemIRLoc loc,
     case SemIR::BuiltinFunctionKind::None:
       CARBON_FATAL("Not a builtin function.");
 
-    case SemIR::BuiltinFunctionKind::PrintInt: {
-      // Providing a constant result would allow eliding the function call.
+    case SemIR::BuiltinFunctionKind::PrintInt:
+    case SemIR::BuiltinFunctionKind::PrintChar:
```
Comment (jonmeow) [inline toolchain/check/eval.cpp]:
```suggestion
    case SemIR::BuiltinFunctionKind::PrintChar:
    case SemIR::BuiltinFunctionKind::PrintInt:
```

File: toolchain/sem_ir/builtin_function_kind.def [hunk 3]
```diff
@@ -17,7 +17,11 @@
 #endif
 
 CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(None)
+
+// Temporary builtins for primitive IO.
 CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(PrintInt)
+CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(PrintChar)
```
Comment (jonmeow) [inline toolchain/sem_ir/builtin_function_kind.def]:
```suggestion
CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(PrintChar)
CARBON_SEM_IR_BUILTIN_FUNCTION_KIND(PrintInt)
```

File: examples/advent2024/day1_part1.carbon [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      Swap(p, from, to - 1);
+      ++from;
+      --to;
+    }
+  }
+  Swap(p, pivot_index, from - 1);
+  return from - 1;
+}
+
+fn Quicksort(p: [i32; 1000]*, from: i32, to: i32) {
```
Comment (jonmeow) [inline examples/advent2024/day1_part1.carbon]: Had you considered moving `Quicksort` out to a library?

File: toolchain/lower/handle_call.cpp [hunk 5]
```diff
@@ -77,19 +77,47 @@ static auto HandleBuiltinCall(FunctionContext& context, SemIR::InstId inst_id,
...
+      llvm::Value* format_string =
+          context.builder().CreateGlobalString("%d\n", "printf.int.format");
+      llvm::Value* arg_value = context.builder().CreateSExtOrTrunc(
+          context.GetValue(arg_ids[0]), i32_type);
+      context.SetLocal(inst_id, context.builder().CreateCall(
+                                    printf, {format_string, arg_value}));
+      return;
+    }
+
+    case SemIR::BuiltinFunctionKind::PrintChar: {
```
Comment (jonmeow) [inline toolchain/lower/handle_call.cpp]: (lexical sort blah blah)

File: examples/advent2024/day1_part1.carbon [hunk 6]
```diff
@@ -0,0 +1,142 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
```
Comment (jonmeow) [inline examples/advent2024/day1_part1.carbon:1]: These files don't really have much in terms of documentation. Would it be worth at least adding a README.md explaining what you're doing here?

**Review by jonmeow — COMMENTED** [review]
Regarding the return type question you just asked -- consider me now neutral, the narrow delta of this change doesn't seem that bad to review. But, maybe fixing global variables should be in the PR title?

File: examples/advent2024/day1_part1.carbon (hunk 4 above)
Comment (zygoloid) [reply examples/advent2024/day1_part1.carbon]: I had, but `import_ref` can't import array types yet, so this is blocked on that.

**Comment by zygoloid** [comment]
Split the bugfix out to #4674.

File: examples/advent2024/day1_part1.carbon (hunk 4 above)
Comment (zygoloid) [reply examples/advent2024/day1_part1.carbon]: #4675 for that.

File: examples/advent2024/day1_part2.carbon (hunk 1 above)
Comment (zygoloid) [reply examples/advent2024/day1_part2.carbon]: Done.

File: examples/advent2024/day1_part1.carbon (hunk 4 above)
Comment (zygoloid) [reply examples/advent2024/day1_part1.carbon]: And now moved out into a library.

File: examples/advent2024/day1_part1.carbon (hunk 6 above)
Comment (zygoloid) [reply examples/advent2024/day1_part1.carbon:1]: Done.

**Review by jonmeow — APPROVED** [review]

**Pull request merged**